### PR TITLE
add proper locking to file_put_contents when using streams

### PR DESCRIPTION
**Pull request by icewind1991**
For #16664

cc @DeepDiver1975 @PVince81

### CODE REVIEW

File: lib/private/files/view.php [hunk 1]
```diff
@@ -535,25 +535,37 @@ public function file_put_contents($path, $data) {
...
+				list($storage, $internalPath) = $this->resolvePath($path);
+				$target = $storage->fopen($internalPath, 'w');
 				if ($target) {
-					list ($count, $result) = \OC_Helper::streamCopy($data, $target);
+					list (, $result) = \OC_Helper::streamCopy($data, $target);
 					fclose($target);
 					fclose($data);
 					$this->updater->update($path);
+
+					$this->unlockFile($path, ILockingProvider::LOCK_EXCLUSIVE);
```
Comment (PVince81) [inline lib/private/files/view.php]: Change back to LOCK_SHARED until the post hooks have fired ?

Comment (icewind1991) [inline lib/private/files/view.php]: Not sure if we want to have a lock around post hooks, there might be valid use cases where a post hooks writes to the file

**Comment by icewind1991** [comment]
[ ] change lock to shared before calling the cache updater to make it work with #16729

**Comment by icewind1991** [comment]
> There was 1 failure:
10:49:17 
10:49:17 1) Test\Encryption\ManagerTest::testGetEncryptionModuleUnknown
10:49:17 Failed asserting that exception message 'Module with id: unknown does not exist.' contains 'Module with id: unknown does not exists.'.

Doesn't sound related...

**Comment by PVince81** [comment]
The post hook might want to do file-related things like creating/updating versions.
I think the look needs to be preserved on the file to avoid people moving away the file while the versions are still being processed.

**Comment by icewind1991** [comment]
@DeepDiver1975 @PVince81 can this be merged?

We can discuss whether or not to lock around post hooks seperatly

**Comment by PVince81** [comment]
Raised post hook discussion here: https://github.com/owncloud/core/issues/16829

:+1: for this PR

**Comment by icewind1991** [comment]
please review @DeepDiver1975 @MorrisJobke

**Comment by DeepDiver1975** [comment]
:+1:

**Pull request merged**